lms/djangoapps/certificates/models.py

```diff
@@ -179,6 +184,322 @@ def certificate_status_for_student(student, course_id):
+    except that:
+
+    1) Example certificates are not associated with a particular user,
+        and are never displayed to students.
+
+    2) We store the "inputs" for generating the example certificate
+        to make it easier to debug when certificate generation fails.
+
+    3) We use dummy values.
+
+    """
+    # Statuses
+    STATUS_STARTED = 'started'
+    STATUS_SUCCESS = 'success'
+    STATUS_ERROR = 'error'
+
+    # Dummy full name for the generated certificate
+    EXAMPLE_FULL_NAME = u'John Doë'
+
+    example_cert_set = models.ForeignKey(ExampleCertificateSet)
+
+    description = models.CharField(
+        max_length=255,
+        help_text=ugettext_lazy(
+            u"A human-readable description of the example certificate.  "
+            u"For example, 'verified' or 'honor' to differentiate between "
+            u"two types of certificates."
+        )
+    )
+
+    # Inputs to certificate generation
+    # We store this for auditing purposes if certificate
+    # generation fails.
+    uuid = models.CharField(
+        max_length=255,
+        default=_make_uuid,
+        db_index=True,
+        unique=True,
+        help_text=ugettext_lazy(
+            u"A unique identifier for the example certificate.  "
+            u"This is used when we receive a response from the queue "
+            u"to determine which example certificate was processed."
+        )
+    )
+
+    access_key = models.CharField(
+        max_length=255,
+        default=_make_uuid,
+        db_index=True,
+        help_text=ugettext_lazy(
+            u"An access key for the example certificate.  "
+            u"This is used when we receive a response from the queue "
+            u"to validate that the sender is the same entity we asked "
+            u"to generate the certificate."
+        )
+    )
+
+    full_name = models.CharField(
+        max_length=255,
+        default=EXAMPLE_FULL_NAME,
+        help_text=ugettext_lazy(u"The full name that will appear on the certificate.")
+    )
+
+    template = models.CharField(
+        max_length=255,
+        help_text=ugettext_lazy(u"The template file to use when generating the certificate.")
+    )
+
+    # Outputs from certificate generation
+    status = models.CharField(
+        max_length=255,
+        default=STATUS_STARTED,
+        choices=(
+            (STATUS_STARTED, 'Started'),
+            (STATUS_SUCCESS, 'Success'),
+            (STATUS_ERROR, 'Error')
+        ),
+        help_text=ugettext_lazy(u"The status of the example certificate.")
+    )
+
+    error_reason = models.TextField(
+        null=True,
+        default=None,
+        help_text=ugettext_lazy(u"The reason an error occurred during certificate generation.")
+    )
+
+    download_url = models.CharField(
+        max_length=255,
+        null=True,
+        default=None,
+        help_text=ugettext_lazy(u"The download URL for the generated certificate.")
+    )
+
```

Comment:
I figured the consistency guarantees from normalizing the tables were more important than optimizing lookup performance of ad-hoc queries.  I'm not anticipating these tables being huge (since they're only used by internal staff), so JOIN'ing these tables shouldn't be prohibitively expensive.